### PR TITLE
staking for direct staked nodes - part 1

### PR DESCRIPTION
**Pull request by ssd04**
## Reasoning behind the pull request
Added staking v4 scenarios:
- Stake funds - happy flow
- unstake funds with deactivation of node if below 2500 -> the rest of funds are distributed as topup at epoch change
- unstake funds with deactivation of node, followed by stake with sufficient ammount does not unstake node at end of epoch

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go [hunk 1]
```diff
@@ -415,3 +418,803 @@ func checkValidatorsRating(t *testing.T, validatorStatistics map[string]*validat
...
+	txStake = generateTransaction(validatorOwner.Bytes, 2, vm.ValidatorSCAddress, stakeValue, txDataField, gasLimitForStakeOperation)
+	stakeTx, err = cs.SendTxAndGenerateBlockTilTxIsExecuted(txStake, maxNumOfBlockToGenerateWhenExecutingTx)
+	require.Nil(t, err)
+	require.NotNil(t, stakeTx)
+
+	err = cs.GenerateBlocks(2) // allow the metachain to finalize the block that contains the staking of the node
+	require.Nil(t, err)
+
+	log.Info("Step 3. Check the stake amount for the owner of the staked nodes")
+	scQuery = &process.SCQuery{
```
Comment (schimih) [inline integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:631]: maybe as separate function? I know that I have at least 2 scenarios using it.

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go [hunk 2]
```diff
@@ -415,3 +418,803 @@ func checkValidatorsRating(t *testing.T, validatorStatistics map[string]*validat
...
+	}
+	result, _, err := metachainNode.GetFacadeHandler().ExecuteSCQuery(scQuery)
+	require.Nil(t, err)
+	require.Equal(t, okReturnCode, result.ReturnCode)
+
+	expectedStaked := big.NewInt(5000)
+	expectedStaked = expectedStaked.Mul(oneEGLD, expectedStaked)
+	require.Equal(t, expectedStaked.String(), string(result.ReturnData[0]))
+
+	log.Info("Step 2. Create from the owner of staked nodes a transaction to unstake 1 EGLD and send it to the network")
```
Comment (schimih) [inline integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:846]: 1 or 10?

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go [hunk 3]
```diff
@@ -415,3 +418,803 @@ func checkValidatorsRating(t *testing.T, validatorStatistics map[string]*validat
...
+		FuncName:   "getTotalStaked",
+		CallerAddr: vm.ValidatorSCAddress,
+		CallValue:  big.NewInt(0),
+		Arguments:  [][]byte{validatorOwner.Bytes},
+	}
+	result, _, err = metachainNode.GetFacadeHandler().ExecuteSCQuery(scQuery)
+	require.Nil(t, err)
+	require.Equal(t, okReturnCode, result.ReturnCode)
+
+	expectedStaked = big.NewInt(4990)
```
Comment (schimih) [inline integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:871]: ok, 10 :)

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go [hunk 4]
```diff
@@ -415,3 +418,803 @@ func checkValidatorsRating(t *testing.T, validatorStatistics map[string]*validat
...
+	stakeTx, err = cs.SendTxAndGenerateBlockTilTxIsExecuted(txStake, maxNumOfBlockToGenerateWhenExecutingTx)
+	require.Nil(t, err)
+	require.NotNil(t, stakeTx)
+
+	err = cs.GenerateBlocks(2) // allow the metachain to finalize the block that contains the staking of the node
+	require.Nil(t, err)
+
+	testBLSKeyStaked(t, cs, metachainNode, blsKeys[1], targetEpoch)
+
+	log.Info("Step 1. Check the stake amount for the owner of the staked nodes")
```
Comment (schimih) [inline integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:1125]: getTotalStaked as separate function?

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go [hunk 5]
```diff
@@ -415,3 +418,803 @@ func checkValidatorsRating(t *testing.T, validatorStatistics map[string]*validat
...
+	txUnStake := generateTransaction(validatorOwner.Bytes, 2, vm.ValidatorSCAddress, zeroValue, txDataField, gasLimitForStakeOperation)
+	unStakeTx, err := cs.SendTxAndGenerateBlockTilTxIsExecuted(txUnStake, maxNumOfBlockToGenerateWhenExecutingTx)
+	require.Nil(t, err)
+	require.NotNil(t, unStakeTx)
+
+	err = cs.GenerateBlocks(2)
+	require.Nil(t, err)
+
+	log.Info("Step 3. Check the outcome of the TX & verify new stake state with vmquery getTotalStaked and getUnStakedTokensList")
+	scQuery = &process.SCQuery{
```
Comment (schimih) [inline integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:1155]: same as previous comments

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go [hunk 6]
```diff
@@ -415,3 +418,803 @@ func checkValidatorsRating(t *testing.T, validatorStatistics map[string]*validat
...
+	txStake = generateTransaction(validatorOwner.Bytes, 3, vm.ValidatorSCAddress, newStakeValue, txDataField, gasLimitForStakeOperation)
+	stakeTx, err = cs.SendTxAndGenerateBlockTilTxIsExecuted(txStake, maxNumOfBlockToGenerateWhenExecutingTx)
+	require.Nil(t, err)
+	require.NotNil(t, stakeTx)
+
+	err = cs.GenerateBlocks(2)
+	require.Nil(t, err)
+
+	log.Info("5. Check the outcome of the TX & verify new stake state with vmquery")
+	scQuery = &process.SCQuery{
```
Comment (schimih) [inline integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:1199]: same as previous comments

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go (hunk 1 above)
Comment (ssd04) [reply integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:631]: right :+1: added separate function

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go [hunk 7]
```diff
@@ -415,3 +418,803 @@ func checkValidatorsRating(t *testing.T, validatorStatistics map[string]*validat
...
+	require.Nil(t, err)
+	require.Equal(t, okReturnCode, result.ReturnCode)
+
+	expectedStaked := big.NewInt(5000)
+	expectedStaked = expectedStaked.Mul(oneEGLD, expectedStaked)
+	require.Equal(t, expectedStaked.String(), string(result.ReturnData[0]))
+
+	log.Info("Step 2. Create from the owner of staked nodes a transaction to unstake 1 EGLD and send it to the network")
+
+	unStakeValue := big.NewInt(10)
```
[the remark this answers is not in the document]
Comment (ssd04) [reply integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:848]: the minimum unstake value is 10, so I have to set it to 10, updated comments and log messages to match

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go (hunk 2 above)
Comment (ssd04) [reply integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:846]: updated comments

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go (hunk 4 above)
Comment (ssd04) [reply integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:1125]: added separate function

**Review by schimih — DISMISSED** [review]

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go [hunk 8]
```diff
@@ -415,3 +418,732 @@ func checkValidatorsRating(t *testing.T, validatorStatistics map[string]*validat
...
+		testChainSimulatorDirectStakedNodesStakingFunds(t, cs, 4)
+	})
+}
+
+func testChainSimulatorDirectStakedNodesStakingFunds(t *testing.T, cs chainSimulatorIntegrationTests.ChainSimulator, targetEpoch int32) {
+	err := cs.GenerateBlocksUntilEpochIsReached(targetEpoch)
+	require.Nil(t, err)
+
+	log.Info("Preconditions. Have an account with 2 staked nodes")
+	privateKey, blsKeys, err := chainSimulator.GenerateBlsPrivateKeys(2)
```
Comment (iulianpascalau) [inline integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:569]: privateKeys *

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go [hunk 9]
```diff
@@ -415,3 +418,732 @@ func checkValidatorsRating(t *testing.T, validatorStatistics map[string]*validat
...
+
+		testChainSimulatorDirectStakedUnstakeFundsWithDeactivation(t, cs, 4)
+	})
+}
+
+func testChainSimulatorDirectStakedUnstakeFundsWithDeactivation(t *testing.T, cs chainSimulatorIntegrationTests.ChainSimulator, targetEpoch int32) {
+	err := cs.GenerateBlocksUntilEpochIsReached(targetEpoch)
+	require.Nil(t, err)
+
+	privateKey, blsKeys, err := chainSimulator.GenerateBlsPrivateKeys(2)
```
Comment (iulianpascalau) [inline integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:790]: privateKeys *

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go [hunk 10]
```diff
@@ -415,3 +418,732 @@ func checkValidatorsRating(t *testing.T, validatorStatistics map[string]*validat
...
+
+		testChainSimulatorDirectStakedUnstakeFundsWithDeactivationAndReactivation(t, cs, 4)
+	})
+}
+
+func testChainSimulatorDirectStakedUnstakeFundsWithDeactivationAndReactivation(t *testing.T, cs chainSimulatorIntegrationTests.ChainSimulator, targetEpoch int32) {
+	err := cs.GenerateBlocksUntilEpochIsReached(targetEpoch)
+	require.Nil(t, err)
+
+	privateKey, blsKeys, err := chainSimulator.GenerateBlsPrivateKeys(2)
```
Comment (iulianpascalau) [inline integrationTests/chainSimulator/staking/stakeAndUnStake_test.go:1065]: privateKeys *

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go (hunk 8 above)
Comment (ssd04) [reply integrationTests/chainSimulator/staking/stakeAndUnStake_test.go]: renamed

File: integrationTests/chainSimulator/staking/stakeAndUnStake_test.go (hunk 9 above)
Comment (ssd04) [reply integrationTests/chainSimulator/staking/stakeAndUnStake_test.go]: renamed

**Review by miiu96 — APPROVED** [review]

**Pull request merged**